### ISSUE TITLE
Updating the API (synchronization of branch master and branch develop)

### ISSUE THREAD

**Issue by adjih** [openschc/openschc#117]
This issue is about the changes to synchronize the  branch `master` with the branch `develop`. It stems from a discussion between openschc contributors.

The main question is related to the API changes, which impact "connectors" or "lower layers" (e.g. [examples/udp](../blob/master/examples/udp), [examples/gateway](../blob/master/examples/gateway/), [examples/simulator](../blob/master/examples/simulator), [examples/scapy](../blob/develop/examples/scapy) , etc. )

In order to make the connectors work with the branch `develop` some changes are necessary. 
Some possibilities are listed in the last section, and feedback is welcome.

### Architecture and API on branch `master`

For reference, the architecture and API of openschc in the `master` branch are "defined" (more or less accurately; the interface with the rule manager is incomplete, and the compression is not called in `master`) in : 
- [src/architecture.py](../blob/master/src/architecture.py)
- [docs/Internals/external-interface.svg](../blob/master/docs/Internals/external-interface.svg) - displayed on this [doc page](https://openschc.github.io/openschc/Internals/Architecture.html)
- (and there exists a data flow [docs/oschc-arch-00.svg](https://raw.githubusercontent.com/openschc/openschc/master/docs/oschc-arch-00.svg) although it is unclear if it is up-to-date with `master`).

### Current changes on branch `develop`

These are some changes between the two branches:

1. there have been changes related to the identifier/address management, i.e. `SCHCProtocol` needs to have both the identifiers (addresses) of the core and the device (which itself is).
    -  `schc_recv` and `schc_send` now take both addresses/identifiers as arguments
    
2. also there is a semantics change in `schc_recv` which returns a packet (when available) for Layer 3 (Upper Layer) instead of letting SCHCProtocol call the upper layer with the packet.

3. as a result of address/identifier management, and proper rule definition, internally, every used address is the in the format of the identifiers of the rule database, e.g. a string such as `"lorawan:70b3d5499126b445"`: 
    - there is the question of under which format the UpperLayer and LowerLayer send and receive those addresses, and how to convert to different formats.
       - in particular: the packet addresses are exchanged in `SCHCProtocol.schc_send`, `SCHCProtocol.schc_recv`, `UpperLayer.recv_packet`, and `LowerLayer.send_packet` (in design `architecture.py`, check in actual code).
    - In addition, for the UpperLayer there is the issue of knowing the address/identifier of the other end based on the IPv6 address.
    
4. there are changes in the scheduler to cancel all events related to one session [src/net_sim_sched.py](../blob/develop/src/net_sim_sched.py), e.g. addition of one method `cancel_session`

### Proposed changes to synchronize both branches

1. for `schc_recv`, `schc_send`: keep both semantics from `master` and `develop`, considering the one from develop as more internal than the ones from branch `master`. 
    - rationale(s):
      - the `schc_recv` semantics (in `master`) of letting SCHCProtocol pass the packet to the UpperLayer is consistent with the `schc_send` semantics of letting SCHCProtocol pass the packets to the LowerLayer ; and is probably expected by the connectors developers. 
      - it makes sense to store one's own address/identifier in SCHCProtocol compared to passing it every time,
    - This can be achieved as follows: keep both pairs (of each branch), by renaming one pair, e.g.:
      - rename `schc_recv`, `schc_send` with branch `master` semantics as `schc_recv_from_lower`, `schc_send_from_upper` 
      - or rename `schc_recv`, `schc_send` of `develop`  to `_schc_recv`, `_schc_send` ;
      - and one version from the other (with one or two changes in the code)
    - Alternately: by renaming `SCHCProtocol` from branch `develop` as `BaseSCHCProtocol` and creating a class `SCHCProtocol` from branch `master` inheriting from `BaseSCHCProtocol`, and implementing the `master` semantics. 

2. for the knowledge of own address/id in the SCHCProtocol (for the version of `master` of `schc_recv`/`schc_send`):
    - this can be implemented in `LowerLayer.get_address`
    - this can also be passed in the constructor of the SCHCProtocol. Rationale: the `position` is also passed to the constructor, so one should consider if this is consistent.

3. for the addresses/identifiers:
    - for the exchanges with the LowerLayer which do need MAC addresses
      - the internal representation of the addresses/identifiers of the rule manager can be consistently used in the openschc API, and helper functions can be provided to do conversions to any format. 
      - (if both semantics of `develop` and `master` are needed it is possible to use inheritance+template methods, or to use object composition to have an optional address conversion object).
    - for the UpperLayer to retrieve the address/identifier of the other end, for some backends, some external step would be needed, as it is only available indirectly inside the rule set. Extra work is needed for this case, and for basic connector cases, a table can be constructed/provided (maybe in the most complex case, provided alongside with the rules).

4. for the `cancel_session`, some possibilities:
    - maybe review the code and implement `cancel_session` inside the SCHC and fragmentation code, not inside the scheduler. Rationale: the semantics of the SCHC sessions are unrelated to the semantics of the scheduler.
    - or rename `cancel_session` to `cancel_event_group` (introducing the equivalent concept of `event group`)
      -  and maybe implement it as an adapter/wrapper to any scheduler. Rationale: there are at least three implementations of the Scheduler interface; one using 'asyncio' loops, one using Python 'sched' module, one "done from scratch", SimulScheduler; and implementing the semantics might require hacking their internals, or more reasonably implementing an adapter/wrapper anyway.
    - if the semantics of canceling session timers was already broken in `master`, this is not an urgent issue.

### Feedback

Your help in selecting the changes would greatly welcomed. Thanks!

**Comment by miri64** [comment]
> there is the question of under which format the UpperLayer and LowerLayer send and receive those addresses, and how to convert to different formats.
> 
>     * in particular: the packet addresses are exchanged in `SCHCProtocol.schc_send`, `SCHCProtocol.schc_recv`, `UpperLayer.recv_packet`, and `LowerLayer.send_packet` (in design `architecture.py`, check in actual code).

IMHO it is for both addresses and data to KISS, and just use python's `bytes` datatype. Most other libraries will provide just that and it is easy to convert any string formats from and to it.

**Comment by ltn22** [comment]
yes, we are not in a coherent approach. Scapy should not be seen in schc_recv and schc_send calls.

User manipulate bytes with schc_send, schc_recv should also returns bytes. The problem is that to effectively send packets on the network we go bask to scapy. One possible way to solve it is to add the iface argument in the schc_recv. When set the packet is sent on the network on that interface. In any cases, the function return None when receving a fragment and the bytes for a compress packet or for the last fragment. False can be used to indicate a wrong RSC.

**Comment by adjih** [comment]
Following the discussion, I propose the following plan:
- For 1.: rename `schc_recv`, `schc_send` with branch master semantics as `schc_recv_from_lower`, `schc_send_from_upper`.
- For 2.: put back the addresses in `LowerLayer` (this will require cosmetic changes to all backends).
- For 3.: for addresses, keep things as they are on branch `develop`, but add an optional object in SCHCProtocol (object composition), to do optional conversion.
- For 4.: implement do-nothing "cancel_session" method on existing connectors which need them, and fix later (it is currently broken anyway).
